### PR TITLE
Feature/wakaama update on new version

### PR DESCRIPTION
**Pull request by moritzholzer**
### Contribution description

This patch updates the wakaama library to a new version (Commit ee0c98da7495e1ab43358b49e970dbc97f73ce3b).
It updates the RIOT code to the new interfaces.

### Testing procedure

Test with the lwm2m example, with leshan demo server.

### CODE REVIEW

**Comment by leandrolanzieri** [comment]
Hey @moritzholzer, thanks for working on this!

For an easier reviewing process, I'd suggest you try to squash unnecessary commits (e.g., the merges). Please, make sure check our commit conventions https://github.com/RIOT-OS/RIOT/blob/master/CONTRIBUTING.md#commit-conventions

Regarding the version bump, do you know whether there are plans from Wakaama on releasing a version 2.0? With this update, which is the highest LwM2M version that is supported?

**Comment by moritzholzer** [comment]
I wil squash the commits later.

The lwm2m version is 1.0, some features 1.1.
See [Which LWM2M version Wakaama Client supports?](https://github.com/eclipse-wakaama/wakaama/issues/729)

From what i can get from the issues in the wakaama repo there is not a fixed plan for a new version but it was discussed in the past without an conclusion.

File: pkg/wakaama/contrib/objects/device.c [hunk 1]
```diff
@@ -47,12 +47,13 @@ static bool reboot;
  * @return COAP_400_BAD_REQUEST         when wrong information has been sent
  * @return COAP_405_METHOD_NOT_ALLOWED  when trying to execute a resource that is not supported
  */
-static uint8_t _execute_cb(uint16_t instance_id, uint16_t resource_id, uint8_t *buffer, int length,
-                           lwm2m_object_t *object);
+static uint8_t _execute_cb(lwm2m_context_t * context, uint16_t instance_id, uint16_t resource_id,
```
Comment (leandrolanzieri) [inline pkg/wakaama/contrib/objects/device.c:50]: Please update the doc

File: pkg/wakaama/contrib/objects/device.c [hunk 2]
```diff
@@ -79,8 +80,8 @@ static uint8_t _read_cb(uint16_t instance_id, int *num_data, lwm2m_data_t **data
  * @return COAP_404_NOT_FOUND               when a resource is not supported
  * @return COAP_500_INTERNAL_SERVER_ERROR   otherwise
  */
-static uint8_t _discover_cb(uint16_t instance_id, int *num_data, lwm2m_data_t **data_array,
-                            lwm2m_object_t *object);
+static uint8_t _discover_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
```
Comment (leandrolanzieri) [inline pkg/wakaama/contrib/objects/device.c:83]: Please update the doc

File: pkg/wakaama/contrib/objects/ipso_sensor_base.c [hunk 3]
```diff
@@ -38,8 +38,8 @@
  * @return COAP_404_NOT_FOUND if the instance was not found
  * @return COAP_500_INTERNAL_SERVER_ERROR otherwise
  */
-static uint8_t _read_cb(uint16_t instance_id, int *num_data, lwm2m_data_t **data_array,
-                        lwm2m_object_t *object);
+static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
```
Comment (leandrolanzieri) [inline pkg/wakaama/contrib/objects/ipso_sensor_base.c:41]: Here and in `_exec_cb` the documentation needs to be updated

File: pkg/wakaama/contrib/objects/light_control.c [hunk 4]
```diff
@@ -56,8 +56,8 @@ typedef struct lwm2m_obj_light_control_inst {
  * @return COAP_404_NOT_FOUND if the instance was not found
  * @return COAP_500_INTERNAL_SERVER_ERROR otherwise
  */
-static uint8_t _read_cb(uint16_t instance_id, int *num_data, lwm2m_data_t **data_array,
-                        lwm2m_object_t *object);
+static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
```
Comment (leandrolanzieri) [inline pkg/wakaama/contrib/objects/light_control.c:59]: Doc on these functions needs updating

File: pkg/wakaama/contrib/objects/security.c [hunk 5]
```diff
@@ -129,8 +129,8 @@ typedef struct lwm2m_obj_security_inst {
  * @retval COAP_404_NOT_FOUND               when resource can't be found
  * @retval COAP_500_INTERNAL_SERVER_ERROR   otherwise
  */
-static uint8_t _read_cb(uint16_t instance_id, int *num_data, lwm2m_data_t *data_array[],
-                        lwm2m_object_t *object);
+static uint8_t _read_cb(lwm2m_context_t * context, uint16_t instance_id, int * num_data,
```
Comment (leandrolanzieri) [inline pkg/wakaama/contrib/objects/security.c:132]: Here as well

**Review by leandrolanzieri — CHANGES_REQUESTED** [review]
Here's a first round of comments

**Comment by moritzholzer** [comment]
I'v worked through the comments.

File: pkg/wakaama/Makefile.dep [hunk 6]
```diff
@@ -40,6 +40,7 @@ endif
 USEMODULE += ztimer
 USEMODULE += ztimer_sec
 USEPKG += tlsf
+USEPKG += random
```
Comment (leandrolanzieri) [inline pkg/wakaama/Makefile.dep:43]:
```suggestion
USEMODULE += random
```

Comment (moritzholzer) [reply pkg/wakaama/Makefile.dep]: done

**Comment by leandrolanzieri** [comment]
I tested it on native and works well

File: pkg/wakaama/Kconfig [hunk 7]
```diff
@@ -27,6 +27,12 @@ config LWM2M_DEVICE_TTL
     help
         Lifetime of the device on the LwM2M server, expressed in seconds.
 
+config LWM2M_COAP_DEFAULT_BLOCK_SIZE
+    int "CoAP block size"
+    default 1024
+    help
+        Block transfer options support only power-of-two block sizes, from 2**4 (16) to 2**10 (1024) bytes.
```
Comment (leandrolanzieri) [inline pkg/wakaama/Kconfig:34]: To constrain this you could either:
- expose only the exponent as a configuration, like so: https://github.com/RIOT-OS/RIOT/blob/faa10032ca7495f5c540eed1c84c49195718572c/drivers/at/Kconfig#L48-L57
- structure it as a `choice`

File: pkg/wakaama/contrib/objects/security.c [hunk 8]
```diff
@@ -718,10 +733,10 @@ static uint8_t _create_cb(uint16_t instance_id, int num_data, lwm2m_data_t *data
     object->instanceList = LWM2M_LIST_ADD(object->instanceList, instance);
 
     /* write incoming data to the instance */
-    result = _write_cb(instance_id, num_data, data_array, object);
+    result = _write_cb(context, instance_id, num_data, data_array, object,LWM2M_WRITE_PARTIAL_UPDATE);
```
Comment (leandrolanzieri) [inline pkg/wakaama/contrib/objects/security.c:736]:
```suggestion
    result = _write_cb(context, instance_id, num_data, data_array, object, LWM2M_WRITE_PARTIAL_UPDATE);
```

**Comment by leandrolanzieri** [comment]
There seems to be a problem with a `printf` on `object_server.c`. Check the CI output

File: pkg/wakaama/Makefile.include [hunk 9]
```diff
@@ -27,5 +19,40 @@ ifneq (,$(or $(CONFIG_LWM2M_WITH_LOGS),$(filter -DCONFIG_LWM2M_WITH_LOGS=1,$(CFL
...
+  endif
+
+  ifeq ($(CONFIG_LWM2M_COAP_DEFAULT_BLOCK_SIZE), 9)
+      COAP_BLOCK_SIZE := 512
+  endif
+
+  ifeq ($(CONFIG_LWM2M_COAP_DEFAULT_BLOCK_SIZE), 10)
+      COAP_BLOCK_SIZE := 1024
+  endif
+  CFLAGS += -DLWM2M_COAP_DEFAULT_BLOCK_SIZE=$(COAP_BLOCK_SIZE)
```
Comment (leandrolanzieri) [inline pkg/wakaama/Makefile.include:54]: Can't we set it as `-DLWM2M_COAP_DEFAULT_BLOCK_SIZE=(1<<$(CONFIG_LWM2M_COAP_DEFAULT_BLOCK_SIZE))` if `CONFIG_LWM2M_COAP_DEFAULT_BLOCK_SIZE` is defined?

Comment (leandrolanzieri) [reply pkg/wakaama/Makefile.include:54]: The shift would be done by the preprocessor, but double-check whether it works well with the -D injection

**Comment by leandrolanzieri** [comment]
You may want to remove the `Makefile.ci` and run `make generate-Makefile.ci` in the example (IT TAKES SOME TIME) to exclude boards that can't fit the application anymore.

File: pkg/wakaama/Makefile.include (hunk 9 above)
Comment (leandrolanzieri) [reply pkg/wakaama/Makefile.include:54]: This seems to at least compile without issues:
```makefile
# NOTE: "Default CoAP block size; Used if not set on a per-target basis"
ifeq (,$(CONFIG_LWM2M_COAP_DEFAULT_BLOCK_SIZE))
  CFLAGS += -DLWM2M_COAP_DEFAULT_BLOCK_SIZE=1024
else
  CFLAGS += -DLWM2M_COAP_DEFAULT_BLOCK_SIZE='1<<$(CONFIG_LWM2M_COAP_DEFAULT_BLOCK_SIZE)'
endif
```

Comment (moritzholzer) [reply pkg/wakaama/Makefile.include]: tested it, it sets the correct value i changed it.

File: pkg/wakaama/Makefile.dep [hunk 10]
```diff
@@ -12,35 +12,12 @@ USEMODULE += wakaama_objects
...
-endif
-
-ifneq (,$(filter wakaama_objects_temperature,$(USEMODULE)))
-    USEMODULE += wakaama_objects_ipso_sensor_base
-endif
-
-ifneq (,$(filter wakaama_objects_voltage,$(USEMODULE)))
-    USEMODULE += wakaama_objects_ipso_sensor_base
-endif
-
```
Comment (leandrolanzieri) [inline pkg/wakaama/Makefile.dep:39]: Why?

Comment (moritzholzer) [reply pkg/wakaama/Makefile.dep:39]: good question i revert it...

**Comment by leandrolanzieri** [comment]
Looks good. There's only one style comment pending. Please squash directly, so that we keep only meaningful commits in the history.

**Comment by moritzholzer** [comment]
Is it enought squashed or are less commits better?

**Comment by leandrolanzieri** [comment]
Thanks @moritzholzer. I think there are some mix ups on the commit history. I see, for example, various changes of the pkg version? (ec5d90f6356a36476898d6df42060dba4d94e2ae and e0fe270c3f4a4ca87eb79f963b90c776b2090d77)

I would suggest having the following commits:
- version bump (updates on pkg makefiles, updates of the patches)
- changes on config and OS services (addition of random seed, Kconfig)
- adaption of contrib to new interface (changes on our object implementations and their docs)
- changes on example application

Also please check the messages. They should always start with the area of changes (e.g., `pkg/wakaama`, `examples/lwm2m`). Also, we use present tense. If you want to elaborate on large changes, do so by leaving a blank line after the short description.

**Comment by moritzholzer** [comment]
Thank you @leandrolanzieri  for the help:) 
I did my best to restructure the commits.

**Comment by leandrolanzieri** [comment]
@moritzholzer there are some whitespace errors (see https://github.com/RIOT-OS/RIOT/actions/runs/11609350032/job/32530338557?pr=20930)

**Comment by moritzholzer** [comment]
I removed the whitespaces.

**Comment by maribu** [comment]
Before the merge queue triggers a timeout, I'll drop and requeue this one.

**Pull request merged**